pkg/ottl/ottlfuncs/func_parse_csv.go

```diff
@@ -0,0 +1,172 @@
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/parseutils"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+const (
+	parseCSVModeStrict       = "strict"
+	parseCSVModeLazyQuotes   = "lazyQuotes"
+	parseCSVModeIgnoreQuotes = "ignoreQuotes"
+)
+
+const (
+	parseCSVDefaultDelimiter = ','
+	parseCSVDefaultMode      = parseCSVModeStrict
+)
+
+type ParseCSVArguments[K any] struct {
+	Target          ottl.StringGetter[K]
+	Header          ottl.StringGetter[K]
+	Delimiter       ottl.Optional[string]
+	HeaderDelimiter ottl.Optional[string]
+	Mode            ottl.Optional[string]
+}
+
+func (p ParseCSVArguments[K]) validate() error {
+	if !p.Delimiter.IsEmpty() {
+		if len([]rune(p.Delimiter.Get())) != 1 {
+			return errors.New("delimiter must be a single character")
+		}
+	}
+
+	if !p.HeaderDelimiter.IsEmpty() {
+		if len([]rune(p.HeaderDelimiter.Get())) != 1 {
+			return errors.New("header_delimiter must be a single character")
+		}
+	}
+
+	return nil
+}
+
+func NewParseCSVFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("ParseCSV", &ParseCSVArguments[K]{}, createParseCSVFunction[K])
+}
+
+func createParseCSVFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*ParseCSVArguments[K])
+	if !ok {
+		return nil, fmt.Errorf("ParseCSVFactory args must be of type *ParseCSVArguments[K]")
+	}
+
+	if err := args.validate(); err != nil {
+		return nil, fmt.Errorf("invalid arguments: %w", err)
+	}
+
+	delimiter := parseCSVDefaultDelimiter
+	if !args.Delimiter.IsEmpty() {
+		delimiter = []rune(args.Delimiter.Get())[0]
+	}
+
+	// headerDelimiter defaults to the chosen delimter,
+	// since in most cases headerDelimiter == delmiter.
+	headerDelimiter := delimiter
+	if !args.HeaderDelimiter.IsEmpty() {
+		headerDelimiter = []rune(args.HeaderDelimiter.Get())[0]
+	}
+
+	mode := parseCSVDefaultMode
+	if !args.Mode.IsEmpty() {
+		mode = args.Mode.Get()
+	}
+
+	switch mode {
+	case parseCSVModeStrict:
+		return parseCSV(args.Target, args.Header, delimiter, headerDelimiter, false), nil
+	case parseCSVModeLazyQuotes:
+		return parseCSV(args.Target, args.Header, delimiter, headerDelimiter, true), nil
+	case parseCSVModeIgnoreQuotes:
+		return parseCSVIgnoreQuotes(args.Target, args.Header, delimiter, headerDelimiter), nil
+	}
+
+	return nil, fmt.Errorf("unknown mode: %s", mode)
+}
+
+func parseCSV[K any](target, header ottl.StringGetter[K], delimiter, headerDelimiter rune, lazyQuotes bool) ottl.ExprFunc[K] {
+	headerDelimiterString := string([]rune{headerDelimiter})
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		targetStr, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, fmt.Errorf("get target: %w", err)
```

Comment:
When we get an error from a getter we can return it directly.

Reply:
Does the error message from `Get` contain the field name? I'd like to add context of whether the get for `target` failed or the get for `header` failed.

Reply:
Thats fair.  If you want to do that, lets make it as human-friendly as possible - `error getting value for target in ParseCSV` or something.

Reply:
Sounds good, modified the error message to your suggestions